content/specs/auth.md

```diff
@@ -0,0 +1,401 @@
+
+The client next makes a Pushed Authorization Request via HTTP POST request. See “Authorization Request” section; some notable details include:
+
+- a randomly generated `state` token is required, and will be used to reference this authorization request with the subsequent response callback
+- PKCE is required, so a secret value is generated and stored, and a derived challenge is included in the request
+- `scope` values are requested here, and must include `atproto`
+- for confidential clients, a `client_assertion` is included, with type `jwt-bearer`, signed using the secret client authentication key
+- the client generates a new DPoP key for the user/session and uses it starting with the PAR request
+- if the auth flow started with an account identifier, the client should pass that starting identifier via the `login_hint` field
+- atproto uses PAR, so the request will be sent as an HTTP POST request to the Authorization Server
+
+The Authorization Server will receive the PAR request and use the `client_id` URL to resolve the client metadata document. The server validates the request and client metadata, then stores information about the session, including binding a DPoP key to the session. The server returns a `request_uri` token to the client, including a DPoP nonce via HTTP header.
+
+The client receives the `request_uri` and prepares to redirect the user. At this point, the client usually needs to persist information about the session to some type of secure storage, so it can be read back after the redirect returns. This might be a database (for a web service backend) or web platform storage like IndexedDB (for a browser app). The client then redirects the user via browser to the Authorization Server’s auth endpoint, including the `request_uri` as a URL parameter.
+
+The Authorization Server uses the `request_uri` to look up the earlier Authorization Request parameters, authenticates the user (which might include sign-in or account selection), and prompts the user with the Authorization Interface. The user might refine any granular requested scopes, then approves or rejects the request. The Authorization Server redirects the user back to the `redirect_uri`, which might be a web callback URL (for web clients), or a native app URI.
+
+The client uses URL query parameters (`state` and `iss`) to look up and verify session information. Using the `code` query parameter, the client then makes an initial token request to the Authorization Server’s token endpoint. The client completes the PKCE flow by including the earlier value in the `code_verifier` field. Confidential clients need to include a client assertion JWT in the token request; see the "Confidential Client" section. The Authorization Server validates the request and returns a set of tokens, as well as a `sub` field indicating the account identifier (DID) for this session, and the `scope` that is covered by the issued access token.
```

Comment:
We should explain that replayed the `code` must be rejected by the AS, and must cause the token previously issued to that `code` to be revoked.

Reply:
The later part does make HTTP glitches/retries brittle during the initial token request, which is too bad. And is it really strictly necessary (mandatory, the sort of thing we will eventually write a compliance test for), or just recommended / best practice?

Reply:
Only `MAY`s and `SHOULD`s in various RFCs (see bellow).

The point of this protection is to ensure that the user realizes it is under attack when it's session gets invalidated. Sadly, there is no way to distinguish an attack from a network error during the token exchange...

When a client fails to exchange a code for a token, it typically *will* implement recovery measures. In case of network error, the recovery method is to initiate a new auth flow, which may be silently accepted by the AS (resulting in no UX degradation). Worst case scenario, the user will have to "accept" again.


https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#section-4.3.1

> When a browser navigates to client.example/redirection_endpoint?code=abcd as a result of a redirect from a provider's authorization endpoint, the URL including the authorization code may end up in the browser's history. An attacker with access to the device could obtain the code and try to replay it.

https://www.rfc-editor.org/rfc/rfc6819.html#section-4.4.1.1

> If an authorization server observes multiple attempts to redeem an authorization "code", the authorization server may want to revoke all tokens granted based on the authorization "code" (see [Section 5.2.1.1](https://www.rfc-editor.org/rfc/rfc6819.html#section-5.2.1.1)).

> The authorization server should enforce a one-time usage restriction (see [Section 5.1.5.4](https://www.rfc-editor.org/rfc/rfc6819.html#section-5.1.5.4)).

https://www.rfc-editor.org/rfc/rfc6819.html#section-5.1.5.4

> For example, if an authorization server observes more than one attempt to redeem an authorization "code", the authorization server may want to revoke all access tokens granted based on the authorization "code" as well as reject the current request.

https://www.rfc-editor.org/rfc/rfc6819.html#section-5.2.3.6

> An authorization server may revoke a client's secret in order to prevent abuse of a revealed secret.
>
> Note: This measure will immediately invalidate any authorization "code" or refresh token issued to the respective client.

Reply:
The issue is worse with refresh_token rotation, as a new auth flow must be initiated, at a time when the user was not already authenticating.

Reply:
I updated text to add this as a "should" in the PKCE session (this is about codes).

tracking discussion of whether this is a "must": https://github.com/bluesky-social/atproto-website/issues/339